src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/ApplicationPropertyPage/ApplicationIconValueProvider.cs

```diff
@@ -0,0 +1,70 @@
+using Microsoft.VisualStudio.IO;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Properties.Package
+{
+    // Sets the ApplicationIcon property to the path to the .ico file. This path is relative
+    // to the project file directory. If the selected file is not within the project directory
+    // tree, the file will be copied into the project file directory. Previously copied files
+    // are not deleted when a new file is selected. ApplicationIcon allows for both full paths
+    // and relative paths anywhere on disk, but we are not utilizing that in this implementation.
+    // The icon file does not need to be included in the project (such as a None/Content item).
+    [ExportInterceptingPropertyValueProvider(ConfigurationGeneral.ApplicationIconProperty, ExportInterceptingPropertyValueProviderFile.ProjectFile)]
+    internal sealed class ApplicationIconValueProvider : InterceptingPropertyValueProviderBase
+    {
+        private readonly UnconfiguredProject _unconfiguredProject;
+        private readonly IFileSystem _fileSystem;
+
+        [ImportingConstructor]
+        public ApplicationIconValueProvider(UnconfiguredProject unconfiguredProject, IFileSystem fileSystem)
+        {
+            _unconfiguredProject = unconfiguredProject;
+            _fileSystem = fileSystem;
+        }
+
+        public override async Task<string?> OnSetPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string>? dimensionalConditions = null)
+        {
+            // Allows the user to clear the application icon.
+            if (string.IsNullOrEmpty(unevaluatedPropertyValue))
+            {
+                return null;
+            }
+
+            // If selected path is invalid, ignore and return previous value.
+            if (!File.Exists(unevaluatedPropertyValue))
+            {
+                return await defaultProperties.GetUnevaluatedPropertyValueAsync(propertyName);
+            }
+
+            // Copy the file into the project directory.
+            // This isn't a limitation of ApplicationIcon; it is to simply mimic the legacy property page interaction.
+            if (_unconfiguredProject.IsOutsideProjectDirectory(unevaluatedPropertyValue))
+            {
+                try
+                {
+                    string fileName = Path.GetFileName(unevaluatedPropertyValue);
+                    var destinationInfo = new FileInfo(Path.Combine(Path.GetDirectoryName(_unconfiguredProject.FullPath), fileName));
+                    if (destinationInfo.Exists && destinationInfo.IsReadOnly)
+                    {
+                        // The file cannot be copied over; return the previous value.
+                        return await defaultProperties.GetUnevaluatedPropertyValueAsync(propertyName);
+                    }
+                    _fileSystem.CopyFile(unevaluatedPropertyValue, destinationInfo.FullName, overwrite: true);
```

Comment:
We should avoid this where possible; this is currently holding a write lock and is going to block the entire solution while this I/O occurs. It also doesn't add the file to the project nor go through source control APIs. I think you should just cut this scenario - the existing behavior is just weird and unexpected.

Reply:
To see these problems; go do this:

1) `<EnableDefaultItems>false</EnableDefaultItems>`
2) Connect the project to TFS source control

Reply:
Compare and contrast with legacy.

Reply:
Ah. I see. It adds it as a `Content` item. I can do that!
![image](https://user-images.githubusercontent.com/17788297/129270134-93c01966-56a0-44ef-afef-1a7c65ec81bd.png)

I talked with the team and it seems somewhat reasonable to block during the I/O since this only occurs when using the UI (prop pages) anyway. For now, it is a better solution than to have this property be broken.

I originally proposed we didn't do the file copy, and instead put a relative path to the file in as the property's value. I even tested it and everything and it worked just fine. The build .exe would pick up the new icon. But, for now, it seems the preferred fix is to keep the functionality the same as it was.

Reply:
I've updated this to add the content item if a previous one didn't exist, or update the content item if it already exists.